modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteServerImpl.java

```diff
@@ -283,26 +291,42 @@ private CompletableFuture<Void> doRestartAsync(IgniteImpl instance) {
 
     @Override
     public CompletableFuture<Void> shutdownAsync() {
+        shutDown = true;
+
         // We don't use attachmentLock here so that users see NodeStoppingException immediately (instead of pausing their operations
         // forever which would happen if the lock was used).
 
         CompletableFuture<Void> result;
 
         synchronized (restartOrShutdownMutex) {
             result = chainRestartOrShutdownAction(this::doShutdownAsync);
-
-            shutDown = true;
         }
 
+        // Trigger stop to avoid situations when a restart future caused a join that cannot complete, so restart never completes,
+        // so our shutdown can never happen.
+        triggerStopOnCurrentIgnite();
+
         return result;
     }
 
+    private void triggerStopOnCurrentIgnite() {
+        IgniteImpl currentIgnite;
+        synchronized (igniteChangeMutex) {
+            currentIgnite = ignite;
+        }
```

Comment:
If that's a "change mutex", does it really make sense to synchronize on it for reading?
It's already volatile, do we just avoid IDE warnings here, is that what's happening?

Reply:
Here, I used it to linearize assignment to `this.ignite` and attempt to invoke `stopAsync()`. As a result:

* either a starting node starts before shutdown executes, so `triggerStopOnCurrentIgnite()` critical section sees the `ignite` that just finished its synchronous start phase and invokes `stopAsync()` on it
* or a starting node finishes its sync start phase after the shutdown gets executed, so `doStartAsync()` sees the `shutdown` flag is `true` and does stop itself

But it's not possible that async start phase finishes and does not see the shutdown, and shutdown does not see the partially-started `this.ignite`.